### PR TITLE
Fix push_universe_context* interfaces to use a consistent ~strict flag

### PR DESCRIPTION
**Pull request by mattam82**
**Kind:** documentation / bug fix

This PR implements a consistent ~strict flag for push_universe_context in environ and safe_typing.
I favored ~strict over ~univpoly because there are actually three behaviors we want to implement: regular monomorphic defs (> Set), (temporarily) template polymorphic inductives (>= Prop during elaboration to find the minimal universes an inductive defs lives in, then > Set) and universe polymorphic ones (>= Set).

- [ ] Entry added to dev/doc/changes.md

### CODE REVIEW

**Comment by ppedrot** [comment]
@mattam82 could the constraints declared by the module system be used so as to generate a soundness bug? Or did it only lead to incompleteness?

**Comment by ppedrot** [comment]
Forgot to mention, but FTR, I need to fiddle with the very same API to safely implement polymorphic sections in the kernel. I guess that this PR should go first then.

**Comment by maximedenes** [comment]
Is this an 8.10 PR?

**Comment by ppedrot** [comment]
@maximedenes Probably yes.

**Comment by vbgl** [comment]
Are there any updates to expect on this PR now that #9918 has been merged?

**Comment by maximedenes** [comment]
@mattam82 it would be nice to get an ETA for this PR, since the current release cycle depends on it. Also, do you need help in order to finish the patch?

**Comment by mattam82** [comment]
I’ll give it a try tomorrow, it’s just two pretty simple commits

**Comment by maximedenes** [comment]
I can do the rebase for you today.

**Comment by mattam82** [comment]
Thanks!

File: kernel/environ.mli [hunk 1]
```diff
@@ -282,7 +282,16 @@ val add_constraints : Univ.Constraint.t -> env -> env
 (** Check constraints are satifiable in the environment. *)
 val check_constraints : Univ.Constraint.t -> env -> bool
 val push_context : ?strict:bool -> Univ.UContext.t -> env -> env
-val push_context_set : ?strict:bool -> Univ.ContextSet.t -> env -> env
+(* [push_context ?(strict=false) ctx env] pushes the universe context to the environment.
+   @raises UGraph.AlreadyDeclared is one of the universes is already declared.
+*)
+val push_context_set : ?enforce_fresh:bool -> ?strict:bool -> Univ.ContextSet.t -> env -> env
+(* [push_context ?enforce_fresh ?(strict=false) ctx env] pushes the universe context to the environment.
```
Comment (ejgallego) [inline kernel/environ.mli]: Doc seems to have a typo in the name of the function.

File: kernel/environ.ml [hunk 2]
```diff
@@ -400,15 +400,19 @@ let add_universes ~lbound ~strict ctx g =
...
 
-let add_universes_set ~lbound ~strict ctx g =
+let add_universes_set ~enforce_fresh ~lbound ~strict ctx g =
   let g = Univ.LSet.fold
             (* Be lenient, module typing reintroduces universes and constraints due to includes *)
-            (fun v g -> try UGraph.add_universe ~lbound ~strict v g with UGraph.AlreadyDeclared -> g)
+            (fun l g -> try UGraph.add_universe ~lbound ~strict l g
+              with UGraph.AlreadyDeclared ->
+                if enforce_fresh then
+                  CErrors.anomaly Pp.(str "Universe " ++ Univ.Level.pr l ++ str" is already declared.")
```
Comment (SkySkimmer) [inline kernel/environ.ml]: I  think we should just let AlreadyDeclared pass through when `enforce_fresh`, there's no point in having the anomaly.
Then you can write it as `try ... with UGraph.AlreadyDeclared when not enforce_fresh -> g`
This matches the behaviour of push_context

**Comment by mattam82** [comment]
@maximedenes yes, let's forget about this one for 8.10

**Comment by mattam82** [comment]
I reworked this so that it only implements the change of interface adding the ~strict label to the argument, making the code clearer and more consistent, without changing any functionality a priori.

I think we'll follow the cleaner path advocated by @ppedrot for template inductives: actually implementing them as proper cumulative polymorphic inductives + the optimization of not storing the universe instance of irrelevant universes.

**Review by ppedrot — APPROVED** [review]
LGTM

**Pull request merged**